test/distributed/test_c10d_object_collectives.py

```diff
@@ -2,139 +2,117 @@
 
 import os
 import sys
-from functools import partial, wraps
+from functools import wraps, partial
 
 import torch
 import torch.distributed as dist
 
-
 if not dist.is_available():
     print("Distributed not available, skipping tests", file=sys.stderr)
     sys.exit(0)
 
-from torch.testing._internal.common_distributed import MultiProcessTestCase, TEST_SKIPS
-from torch.testing._internal.common_utils import run_tests, TEST_WITH_DEV_DBG_ASAN
+from torch.testing._internal.common_device_type import instantiate_device_type_tests
+from torch.testing._internal.common_distributed import (
+    MultiProcessTestCase,
+    skip_if_lt_x_gpu,
+    DistributedTestBase,
+    TEST_SKIPS,
+)
+
 
+from torch.testing._internal.common_utils import (
+    run_tests,
+    TEST_WITH_DEV_DBG_ASAN,
+    TEST_CUDA,
+    TEST_HPU,
+    skipIfHpu
+)
 
 if TEST_WITH_DEV_DBG_ASAN:
-    print(
-        "Skip dev-asan as torch + multiprocessing spawn have known issues",
-        file=sys.stderr,
-    )
+    print("Skip dev-asan as torch + multiprocessing spawn have known issues", file=sys.stderr)
     sys.exit(0)
 
-BACKEND = dist.Backend.NCCL if torch.cuda.is_available() else dist.Backend.GLOO
+if TEST_HPU:
+    BACKEND = dist.Backend.HCCL
+    device_count = torch.hpu.device_count()
+    DEVICE = "hpu"
+elif TEST_CUDA:
+    BACKEND = dist.Backend.NCCL 
+    device_count = torch.cuda.device_count()
+    DEVICE = "cuda"
+else :
+    BACKEND = dist.Backend.GLOO
+    DEVICE = "cpu"
+
+
+WORLD_SIZE = min(4, max(2, device_count))
 
 
 def with_comms(func=None):
```

Comment:
Can we please properly type annotation this function. Decorators are important to type properly as they lose all typing info of things they decorate otherwise.e
```suggestion
def with_comms(func=None):